File: 15 09 25/nqueens.swift
class Solution {
    func solveNQueens(_ n: Int) -> [[String]] {
        var res = [[String]]()
        var board = Array(repeating: Array(repeating: ".", count: n), count: n)
        
        func backtrack(_ row: Int, _ cols: inout Set<Int>, _ diag1: inout Set<Int>, _ diag2: inout Set<Int>) {
            if row == n {
                res.append(board.map { $0.joined() })
                return
            }
            
            for c in 0..<n {
                if cols.contains(c) || diag1.contains(row - c) || diag2.contains(row + c) {
                    continue
                }
                board[row][c] = "Q"
                cols.insert(c)
                diag1.insert(row - c)
                diag2.insert(row + c)
                
                backtrack(row + 1, &cols, &diag1, &diag2)
                
                board[row][c] = "."
                cols.remove(c)
                diag1.remove(row - c)
                diag2.remove(row + c)
            }
        }
        
        var cols = Set<Int>()
        var diag1 = Set<Int>()
        var diag2 = Set<Int>()
        backtrack(0, &cols, &diag1, &diag2)
        return res
    }
}

File: 15 09 25/nqueens.py
class Solution:
    def solveNQueens(self, n: int):
        res = []
        board = [["."] * n for _ in range(n)]
        cols = set()
        diag1 = set()  # r - c
        diag2 = set()  # r + c

        def backtrack(r):
            if r == n:
                res.append(["".join(row) for row in board])
                return
            for c in range(n):
                if c in cols or (r-c) in diag1 or (r+c) in diag2:
                    continue
                board[r][c] = "Q"
                cols.add(c); diag1.add(r-c); diag2.add(r+c)
                backtrack(r+1)
                board[r][c] = "."
                cols.remove(c); diag1.remove(r-c); diag2.remove(r+c)

        backtrack(0)
        return res

    def solveNQueensRecursive(self, n):
        def f(r=0, cols=set(), d1=set(), d2=set(), board=[]):
            if r==n: return [["".join(row) for row in board]]
            return [res for c in range(n) if c not in cols and r-c not in d1 and r+c not in d2
                    for res in f(r+1, cols|{c}, d1|{r-c}, d2|{r+c}, board+[["."]*c+["Q"]+["."]*(n-c-1)])]
        return f()

